include/nbl/builtin/hlsl/emulated/emulated_float64_t.hlsl

```diff
@@ -1,749 +1,788 @@
-
-            const uint32_t toBitSize = sizeof(To) * 8;
-            const AsUint sign = AsUint(ieee754::extractSign(v.data) << (toBitSize - 1));
-            const AsUint biasedExponent = AsUint(exponent + ieee754::traits<To>::exponentBias) << ieee754::traits<To>::mantissaBitCnt;
-            const AsUint mantissa = AsUint(v.data >> (ieee754::traits<float64_t>::mantissaBitCnt - ieee754::traits<To>::mantissaBitCnt)) & ieee754::traits<To>::mantissaMask;
-
-            return bit_cast<To>(sign | biasedExponent | mantissa);
-        }
-
-        // NOTE: casting from negative float to unsigned int is an UB, function will return abs value in this case
-        if (is_integral<To>::value)
-        {
-            const int exponent = ieee754::extractExponent(v.data);
-            if (exponent < 0)
-                return 0;
-
-            uint64_t unsignedOutput = ieee754::extractMantissa(v.data) & 1ull << ieee754::traits<float64_t>::mantissaBitCnt;
-            const int shiftAmount = exponent - int(ieee754::traits<float64_t>::mantissaBitCnt);
-
-            if (shiftAmount < 0)
-                unsignedOutput <<= -shiftAmount;
-            else
-                unsignedOutput >>= shiftAmount;
-
-            if (is_signed<To>::value)
-            {
-                int64_t signedOutput64 = unsignedOutput & ((1ull << 63) - 1);
-                To signedOutput = To(signedOutput64);
-                if (ieee754::extractSignPreserveBitPattern(v.data) != 0)
-                    signedOutput = -signedOutput;
-
-                return signedOutput;
-            }
-
-            return To(unsignedOutput);
-        }
-
-        // assert(false);
-        return 0xdeadbeefbadcaffeull;
-    }
-};
-
-}
-
-DEFINE_BIT_CAST_SPEC(emulated_float64_t<true, true>);
-DEFINE_BIT_CAST_SPEC(emulated_float64_t<false, false>);
-DEFINE_BIT_CAST_SPEC(emulated_float64_t<true, false>);
-DEFINE_BIT_CAST_SPEC(emulated_float64_t<false, true>);
-
-//template<bool FastMath, bool FlushDenormToZero>
-//struct is_floating_point<emulated_float64_t<FastMath, FlushDenormToZero> > : bool_constant<true> {};
-
-namespace ieee754
-{
-IMPLEMENT_IEEE754_FUNC_SPEC_FOR_EMULATED_F64_TYPE(emulated_float64_t<true, true>);
-IMPLEMENT_IEEE754_FUNC_SPEC_FOR_EMULATED_F64_TYPE(emulated_float64_t<false, false>);
-IMPLEMENT_IEEE754_FUNC_SPEC_FOR_EMULATED_F64_TYPE(emulated_float64_t<true, false>);
-IMPLEMENT_IEEE754_FUNC_SPEC_FOR_EMULATED_F64_TYPE(emulated_float64_t<false, true>);
-}
-
-}
-}
-
-#undef FLOAT_ROUND_NEAREST_EVEN
-#undef FLOAT_ROUND_TO_ZERO
-#undef FLOAT_ROUND_DOWN
-#undef FLOAT_ROUND_UP
-#undef FLOAT_ROUNDING_MODE
-
-#undef IMPLEMENT_IEEE754_FUNC_SPEC_FOR_EMULATED_F64_TYPE
-#undef DEFINE_BIT_CAST_SPEC
-#undef DEFINE_EMULATED_FLOAT64_STATIC_CAST
-
-#endif
+#ifndef _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_INCLUDED_
+#define _NBL_BUILTIN_HLSL_EMULATED_FLOAT64_T_INCLUDED_
+
+#include <nbl/builtin/hlsl/emulated/emulated_float64_t_impl.hlsl>
```

Comment:
I'd probably drop the `emulated_` filename prefix, if its already in `emulated` folder

Reply:
"emulated" namespace then?